app/scripts/migrations/088.test.ts

```diff
@@ -724,7 +725,12 @@ describe('migration #88', () => {
     expect(newStorage.data).toStrictEqual(oldData);
   });
 
-  it('captures an exception if it has no TokenListController property', async () => {
+  it('Logs a warning if it has no TokenListController property', async () => {
```

Comment:
In the other branches we hadn't bothered testing the warnings, since they're just meant for debugging purposes and it's less important to maintain this functionality over time. So I think it would be OK to delete these tests.

If we did want to test them though, I'd recommend finding an alternative approach for mocking here. At the moment if one test fails, it can leave `console.warn` mocked causing confusion for later tests (which would be very confusing if you investigate test failures in your console output from bottom-to-top)

We want any mocks to be automatically restored in any circumstances, success or failure. The easiest way to do that is probably to use Jest to apply the mock, because they we can rely on `restoreMocks`/`resetMocks` to reset it for us (we haven't enabled `resetMocks` in this repo yet, but `resetAllMocks` is already manually called in `beforeEach` in this test suite so we're covered here).

We can do this using the Jest `spyOn` method: https://jestjs.io/docs/jest-object#jestspyonobject-methodname

Contrary to what the name might suggest, it's not just for spying. `spyOn` returns a mock function that you can give a mock implementation as well, just like we can for `jest.fn()`. Though in this case, it seems that the spying is all we need anyway. Using `spyOn` here rather than setting `console.warn` directly to the mock function removes the need for the tear down phase.

In more complex cases like this, where we need a tear down phase and the built-in functionality of Jest isn't sufficient, I'd recommend the `with___` helper function pattern that Elliot describes here: https://github.com/MetaMask/contributor-docs/blob/add-more-testing-guidelines/guides/unit-testing.md#avoid-beforeeach-and-aftereach
We've used it extensively in the network controller tests to abstract away setup and teardown and ensure it runs reliably even upon failure. But that would be overkill here.

Reply:
Thanks for the detailed explanation, Mark. I've switched to using `spyOn` and noted the `with__` pattern you linked for future use.